src/oop/special_class_method/hard/q7/MainTest.java

```diff
@@ -0,0 +1,260 @@
+                {value2, int.class}
+        };
+        Object classInstance = testClass.createInstance(arguments);
+        String incorrectDoubleValueMessage =
+                "Your " + className + " constructor does not initialize the " + doubleFieldName + " attribute to the correct value.";
+        String incorrectStringValueMessage =
+                "Your " + className + " constructor does not initialize the " + intFieldName + " attribute to the correct value.";
+        _assertEquals(value1, testClass.getFieldValue(classInstance, doubleFieldName),
+                incorrectDoubleValueMessage);
+        _assertEquals(value2, testClass.getFieldValue(classInstance, intFieldName),
+                incorrectStringValueMessage);
+    }
+
+    @Test
+    public void piggyBankClassHasCorrectConstructor2() throws Throwable {
+        Class<?>[] constructorClasses = {double.class, int.class};
+        Object classInstance = testClass.createInstance();
+        String incorrectDoubleValueMessage =
+                "Your " + className + " constructor does not initialize the " + doubleFieldName + " attribute to the correct value.";
+        String incorrectStringValueMessage =
+                "Your " + className + " constructor does not initialize the " + intFieldName + " attribute to the correct value.";
+        String incorrectConstructorVisibilityModifier =
+                "Your " + className + " constructor does not have the correct visibility modifier.";
+        _assertEquals(0.0, testClass.getFieldValue(classInstance, doubleFieldName),
+                incorrectDoubleValueMessage);
+        assertEquals(0, testClass.getFieldValue(classInstance, intFieldName),
+                incorrectStringValueMessage);
+        assertTrue(testClass.hasModifier(constructorClasses, "public"), incorrectConstructorVisibilityModifier);
+    }
+
+    private static Stream<Arguments> doubleInputProvider() {
+        return Stream.of(
+                Arguments.of(4535.4, 432.2),
+                Arguments.of(234.4, 655.3),
+                Arguments.of(654.45, 453.3),
+                Arguments.of(43.3, 654.2)
+
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("doubleInputProvider")
+    public void piggyBankClassHasCorrectSetMoneyMethod(double initialValue, double setValue) throws Throwable {
+        Object[][] arguments = {
+                {initialValue, double.class},
+                {0, int.class}
+        };
+        Object classInstance = testClass.createInstance(arguments);
+        Object[][] setSizeArguments = {
+                {setValue, double.class}
+        };
+        String[] setMethodModifiers = {"public"};
+        String incorrectSetMethodMessage = String.join(" ",
+                "Your", setDoubleMethodName, "method does not correctly set the value of the", doubleFieldName, "attribute.");
+        Object setMethodOutput = testClass.callMethod(setDoubleMethodName, setSizeArguments, setMethodModifiers, classInstance);
```

Comment:
use .hasMethod first with void.type instead of assertNull here and everywhere else